backend/molgenis-emx2-webapi/src/main/java/org/molgenis/emx2/web/MessageApi.java

```diff
@@ -0,0 +1,121 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import spark.Request;
+import spark.Response;
+
+public class MessageApi {
+
+  private static final Logger logger = LoggerFactory.getLogger(MessageApi.class);
+  private static final Parser parser = new Parser();
+
+  public static String send(Request request, Response response) {
+    logger.info("Received message request");
+    Schema schema = MolgenisWebservice.getSchema(request);
+    if (schema == null) {
+      throw new MolgenisException("Cannot handle send action, schema is null");
+    }
+
+    ObjectMapper objectMapper = new ObjectMapper();
+    SendMessageAction sendMessageAction;
+    Map<String, Object> validationFilter;
+    try {
+      sendMessageAction = objectMapper.readValue(request.body(), SendMessageAction.class);
+      validationFilter = objectMapper.readValue(sendMessageAction.recipientsFilter(), Map.class);
+    } catch (JsonProcessingException e) {
+      response.status(422);
+      return "Error parsing request: " + e.getMessage();
+    }
+
+    String recipientsQuery = schema.getSettingValue(Constants.CONTACT_RECIPIENTS_QUERY_SETTING_KEY);
+
+    try {
+      parser.parseDocument(recipientsQuery);
+    } catch (Exception e) {
+      throw new MolgenisException(
+          "Cannot handle send action, schema setting "
+              + Constants.CONTACT_RECIPIENTS_QUERY_SETTING_KEY
+              + " is not a valid graphql query");
+    }
+
+    MolgenisSession session = sessionManager.getSession(request);
+    GraphQL gql = session.getGraphqlForSchema(schema.getName());
+
+    final ExecutionResult executionResult =
+        gql.execute(ExecutionInput.newExecutionInput(recipientsQuery).variables(validationFilter));
+    if (!executionResult.getErrors().isEmpty()) {
+      response.status(500);
+      return "Error parsing request: " + executionResult.getErrors().get(0).getMessage();
+    }
+    Map<String, Object> resultMap = executionResult.toSpecification();
+
+    final List<String> recipients = new java.util.ArrayList<>(Collections.emptyList());
+    try {
+      recipients.addAll(EmailValidator.validationResponseToRecievers(resultMap));
+    } catch (Exception e) {
+      response.status(500);
+      return "Error parsing request: " + e.getMessage();
+    }
+
+    if (recipients.isEmpty()) {
+      response.status(400);
+      return "No recipients found for given filter";
+    }
+
+    // send email to all recipients on allow list
+    EmailSettings settings = loadEmailSettings(schema);
+    EmailService emailService = new EmailService(settings);
+
+    final Boolean sendResult =
+        emailService.send(recipients, sendMessageAction.subject(), sendMessageAction.body());
+    return sendResult.toString();
+  }
+
+  private static EmailSettings loadEmailSettings(Schema schema) {
+    EmailSettings.EmailSettingsBuilder builder = new EmailSettings.EmailSettingsBuilder();
+
+    setEmailSetting(schema, "EMAIL_HOST", builder::host);
```

Comment:
for now okay, but I kinda feel the need to convert this into one setting of type json and then have validation on that json schema. And if we play it smartly we could use a emx2 metadata schema to manage such json object using standard RowEdit user interface forms (as if it was a database row)?

Reply:
thought about doing this initially , but think using simple key val fields  is probably a better solution, if we want we can add a simple email setting tab to the ui and configure a form to show inputs ( and add validation and stuff). why a user type json and make "" error when we can have them add a string ( of add a form )